File: ocaml/lib/categorical_combinators.ml
type 'a term =
  | Var of int
  | App of 'a term * 'a term
  | Lam of string * 'a term

let rec eval (t: 'a term) (env: (string * 'a) list): 'a =
  match t with
    | Var (n) -> List.nth env n
    | App (t1, t2) ->
      let v1 = eval t1 env in
      let v2 = eval t2 env in
      v1 (v2)
    | Lam (x, t) ->
      fun v -> eval t ((x, v) :: env)

let fix (t: 'a term -> 'a term) : 'a term =
  let x = Var 0 in
  Lam ("x", App (t x, x))

let id = Lam ("x", Var 0)

let k = Lam ("x", Lam ("y", x))

let b = Lam ("x", Lam ("y", Lam ("z", y)))

File: python/categorical_combinators.py
from typing import Any, Callable

class Term:
  def __init__(self, value):
    self.value = value

  def eval(self, env: dict[str, Any]) -> Any:
    match self.value:
      case int(n):
        return env[str(n)]
      case (f, arg):
        return f.eval(env)(arg.eval(env))
      case (var, body):
        new_env = {**env, var: arg}
        return body.eval(new_env)

def fix(f: Callable[[Term], Term]) -> Term:
  return Term((lambda x: f(Term((x, Term(x))))))

def id() -> Callable[[Any], Any]:
  return lambda x: x

def k() -> Callable[[Any], Callable[[Any], Any]]:
  return lambda x: lambda y: x

def b() -> Callable[[Any], Callable[[Any], Callable[[Any], Any]]]:
  return lambda x: lambda y: lambda z: y